src/com/amazon/ion/impl/ReaderLookaheadBuffer.java

```diff
@@ -0,0 +1,48 @@
+package com.amazon.ion.impl;
+
+import com.amazon.ion.BufferEventHandler;
+import com.amazon.ion.IonReader;
+import com.amazon.ion.system.IonReaderBuilder;
+
+/**
+ * Interface for lookahead buffers that enable incremental reading of streaming data.
+ */
+public interface ReaderLookaheadBuffer {
+
+    /**
+     * If possible, fills the input pipe with enough bytes to enable one more successful call to
+     * {@link IonReader#next()} on the non-incremental reader attached to the pipe. If not enough bytes were available
+     * in the raw input stream to complete the next top-level user value, calling {@link #moreDataRequired()} after this
+     * method returns will return `true`. In this case, this method must be called again before calling
+     * `IonReader.next()` to position the reader on this value. Otherwise, `moreDataRequired()` will return `false` and
+     * a call to `IonReader.next()` may be made to position the reader on this value. Implementations may throw
+     * `IonException` if invalid Ion data is detected. Any exceptions thrown when invoking {@link BufferEventHandler}
+     * methods will be propagated. Implementations may define additional exceptional cases.
+     * @throws Exception if thrown by a handler method or if an IOException is thrown by the underlying InputStream.
+     */
```

Comment:
I agree that the comment about BufferEventHandler should be moved to ReaderLookaheadBufferBase. Done.

I think the IonReader references should remain, however; this interface is inextricably linked to IonReader (hence the `newIonReader`method).

Regarding why both the ReaderLookaheadBuffer interface and ReaderLookaheadBufferBase abstract class both exist, that's a good callout. It's because I separately implemented some incremental reader support for text that I'm not proposing in this PR, but may later. I ended up defining a class that extends ReaderLookaheadBufferBase capable of incrementally reading newline-delimited Ion text. I also defined a "no-op lookahead buffer" that trivially implements the ReaderLookaheadBuffer and allows the user to use the same underlying machinery for blocking use cases. That satisfied me that the current factoring probably has some value, but I'm open to suggestions.